### PR TITLE
feat: support provider aliases

### PR DESCRIPTION
**Pull request by suzuki-shunsuke**
https://github.com/terraform-linters/tflint-ruleset-aws/issues/331

Follow up https://github.com/terraform-linters/tflint-ruleset-aws/pull/332#issuecomment-1118514083

### CODE REVIEW

**Comment by wata727** [comment]
Umm... As you say, it seems that it is necessary to reimplement the `decodeProviderConfigRef`.
https://github.com/hashicorp/terraform/blob/3fbedf25430ead97eb42575d344427db3c32d524/internal/configs/resource.go#L498

I think we doesn't need to reproduce all implementations completely, for example, support for legacy configuration language can be ignored. The important thing is to get the list of traversals with `hcl.AbsTraversalForExpr` and get the root name and step name.

**Comment by suzuki-shunsuke** [comment]
c3caa6c Copied required code.
It seems to work well.
I guess we have to take the original license into account.

**Comment by suzuki-shunsuke** [comment]
Mozilla Public License, version 2.0
https://github.com/hashicorp/terraform/blob/main/LICENSE

File: aws/runner.go [hunk 1]
```diff
@@ -10,31 +13,58 @@ import (
...
-
-		client, err = NewClient(config.toCredentials().Merge(credentials))
-		if err != nil {
-			return nil, err
+		for k, cred := range credentials {
+			client, err := NewClient(config.toCredentials().Merge(cred))
+			if err != nil {
+				return nil, err
+			}
+			clients[k] = client
```
Comment (wata727) [inline aws/runner.go:34]: I noticed that the client is not initialized as before if you don't declare any provider blocks. Previously, if you wrote a credential in `.tflint.hcl`, you should have been able to enable deep checking without a provider block. e.g.

```hcl
plugin "aws" {
  enabled = true
  access_key = "AWS_ACCESS_KEY",
  secret_key = "AWS_SECRET_KEY",
}
```

```hcl
resource "aws_instance" "foo" {
  ami = "invalid"
}
```

File: rules/api/aws_instance_invalid_ami.go [hunk 2]
```diff
@@ -72,13 +72,18 @@ func (r *AwsInstanceInvalidAMIRule) Check(rr tflint.Runner) error {
 			continue
 		}
 
+		awsClient, err := runner.AwsClient(resource.Body.Attributes)
```
Comment (wata727) [inline rules/api/aws_instance_invalid_ami.go:75]: You need to get the provider attribute explicitly like any other rule.

File: rules/api/aws_launch_configuration_invalid_image_id.go [hunk 3]
```diff
@@ -72,13 +72,18 @@ func (r *AwsLaunchConfigurationInvalidImageIDRule) Check(rr tflint.Runner) error
 			continue
 		}
 
+		awsClient, err := runner.AwsClient(resource.Body.Attributes)
```
Comment (wata727) [inline rules/api/aws_launch_configuration_invalid_image_id.go:75]: Same as aws_instance_invalid_rule.

**Comment by wata727** [comment]
> I guess we have to take the original code's license into account.

Yes. We changed TFLint's license 4 years ago to follow this case. MPL 2.0 is a weak copyleft and the Large Work is allowed under the same license.
https://github.com/terraform-linters/tflint/pull/245

File: rules/api/aws_launch_configuration_invalid_image_id.go (hunk 3 above)
Comment (suzuki-shunsuke) [reply rules/api/aws_launch_configuration_invalid_image_id.go:78]: Fixed. 821423f

File: rules/api/aws_instance_invalid_ami.go (hunk 2 above)
Comment (suzuki-shunsuke) [reply rules/api/aws_instance_invalid_ami.go:78]: Fixed. 821423f

File: aws/runner.go (hunk 1 above)
Comment (suzuki-shunsuke) [reply aws/runner.go:34]: f73855a2e4d5ab667d848e10eaa0c559f47febfa

Comment (suzuki-shunsuke) [reply aws/runner.go:34]: https://github.com/terraform-linters/tflint-ruleset-aws/blob/master/docs/deep_checking.md#shared-credentials

> If these configurations are defined in the provider block, they will also be taken into account. But the priority is lower than the above way.

Hmm... Maybe we have to fix f73855a according to the priority.

Comment (suzuki-shunsuke) [reply aws/runner.go:34]: The priority of `plugin` setting is higher than `provider`.

Comment (suzuki-shunsuke) [reply aws/runner.go:37]: Fixed. 313f0da9624d402e23056bdba9b990fe202e1cd8

**Review by wata727 — APPROVED** [review]
🎉

**Comment by suzuki-shunsuke** [comment]
Thank you for your support!

**Pull request merged**